File: library/src/main/java/com/coolerfall/download/DownloadManager.java
package com.coolerfall.download;

import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Context;
import android.net.Uri;
import android.os.Environment;
import android.provider.MediaStore;
import android.webkit.MimeTypeMap;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;

import static android.os.Environment.DIRECTORY_DOWNLOADS;
import static com.coolerfall.download.Preconditions.checkNotNull;
import static com.coolerfall.download.Utils.copy;
import static com.coolerfall.download.Utils.createDefaultDownloader;
import static com.coolerfall.download.Utils.resolvePath;

/**
 * A manager used to manage the downloading.
 *
 * @author Vincent Cheung (coolingfall@gmail.com)
 */
public final class DownloadManager {
  private final Context context;
  private final Downloader downloader;
  private final int threadPoolSize;
  private final Logger logger;
  private DownloadRequestQueue downloadRequestQueue;
  private final String rootDownloadDir;

  DownloadManager(Builder builder) {
    context = checkNotNull(builder.context, "context == null").getApplicationContext();
    downloader = checkNotNull(builder.downloader, "downloader == null");
    threadPoolSize = builder.threadPoolSize;
    logger = builder.logger;
    downloadRequestQueue = new DownloadRequestQueue(threadPoolSize, logger);
    downloadRequestQueue.start();

    File downlodDir = context.getExternalFilesDir(DIRECTORY_DOWNLOADS);
    checkNotNull(downlodDir, "shared storage is not currently available");
    rootDownloadDir = downlodDir.getAbsolutePath();
  }

  /**
   * Add one download request into the queue.
   *
   * @param request download request
   * @return download id, if the id is not set, then manager will generate one.
   * if the request is in downloading, then -1 will be returned
   */
  public int add(DownloadRequest request) {
    checkNotNull(request, "request == null");
    if (isDownloading(request.uri().toString())) {
      return -1;
    }

    request.context(context);
    request.rootDownloadDir(rootDownloadDir);
    request.downloader(downloader.copy());

    /* add download request into download request queue */
    return downloadRequestQueue.add(request) ? request.downloadId() : -1;
  }

  /**
   * Query download from download request queue.
   *
   * @param downloadId download id
   * @return download state
   */
  public DownloadState query(int downloadId) {
    return downloadRequestQueue.query(downloadId);
  }

  /**
   * Query download from download request queue.
   *
   * @param url download url
   * @return download state
   */
  DownloadState query(String url) {
    return downloadRequestQueue.query(Uri.parse(url));
  }

  /**
   * To check if the download was in the request queue.
   *
   * @param downloadId downalod id
   * @return true if was downloading, otherwise return false
   */
  public boolean isDownloading(int downloadId) {
    return query(downloadId) != DownloadState.INVALID;
  }

  /**
   * To check if the download was in the request queue.
   *
   * @param url downalod url
   * @return true if was downloading, otherwise return false
   */
  public boolean isDownloading(String url) {
    return query(url) != DownloadState.INVALID;
  }

  /**
   * Get the download task size.
   *
   * @return the task size
   */
  public int getTaskSize() {
    return downloadRequestQueue == null ? 0 : downloadRequestQueue.getDownloadingSize();
  }

  /**
   * Cancel the download according to download id.
   *
   * @param downloadId download id
   * @return true if download has canceled, otherwise return false
   */
  public boolean cancel(int downloadId) {
    return downloadRequestQueue.cancel(downloadId);
  }

  /**
   * Cancel all the downloading in queue.
   */
  public void cancelAll() {
    downloadRequestQueue.cancelAll();
  }

  /**
   * Release all the resource.
   */
  public void release() {
    if (downloadRequestQueue != null) {
      downloadRequestQueue.release();
      downloadRequestQueue = null;
    }
  }

  /**
   * Copy downloaded file to external public download directory.
   *
   * @param filepath filepath of downloaded file
   * @return true if copy successfully, otherwise return false
   */
  public boolean copyToPublicDownloadDir(String filepath) {
    File publicDownloadDir =
        Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS);
    String dir = publicDownloadDir.getAbsolutePath();
    if (!filepath.startsWith(rootDownloadDir)) {
      throw new IllegalArgumentException("Only files of current app can be exported");
    }
    String filename = filepath.substring(filepath.lastIndexOf(File.separator) + 1);
    String outputFilepath = resolvePath(dir, filepath.substring(rootDownloadDir.length() + 1));
    FileInputStream fis;
    try (OutputStream os = openOutputStream(outputFilepath, filename)) {
      fis = new FileInputStream(filepath);
      copy(fis, os);
      return true;
    } catch (Exception e) {
      logger.log("Failed to copy file to public download directory: " + e.getMessage());
      return false;
    }
  }

  private OutputStream openOutputStream(String filepath, String filename) throws IOException {
    if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.Q) {
      ContentValues contentValues = new ContentValues();
      contentValues.put(MediaStore.Files.FileColumns.RELATIVE_PATH,
          Environment.DIRECTORY_DOWNLOADS);
      contentValues.put(MediaStore.Files.FileColumns.DISPLAY_NAME, filename);
      int index = filename.lastIndexOf(".");
      if (index > 0) {
        String mimeType =
            MimeTypeMap.getSingleton().getMimeTypeFromExtension(filename.substring(index + 1));
        contentValues.put(MediaStore.Files.FileColumns.MIME_TYPE, mimeType);
      }
      ContentResolver contentResolver = context.getContentResolver();
      Uri uri = contentResolver.insert(MediaStore.Downloads.EXTERNAL_CONTENT_URI, contentValues);
      if (uri == null) {
        throw new IOException("Cannot get shared download directory");
      }
      return contentResolver.openOutputStream(uri);
    } else {
      return new FileOutputStream(filepath);
    }
  }

  public Builder newBuilder() {
    return new Builder(this);
  }

  public static final class Builder {
    private Context context;
    private Downloader downloader;
    private int threadPoolSize;
    private Logger logger;

    public Builder() {
      this.threadPoolSize = 3;
      this.logger = Logger.EMPTY;
    }

    Builder(DownloadManager downloadManager) {
      this.context = downloadManager.context;
      this.downloader = downloadManager.downloader;
      this.threadPoolSize = downloadManager.threadPoolSize;
      this.logger = downloadManager.logger;
    }

    public Builder context(Context context) {
      this.context = context;
      return this;
    }

    public Builder downloader(Downloader downloader) {
      this.downloader = downloader;
      return this;
    }

    public Builder threadPoolSize(int threadPoolSize) {
      this.threadPoolSize = threadPoolSize;
      return this;
    }

    public Builder logger(Logger logger) {
      this.logger = logger;
      return this;
    }

    public DownloadManager build() {
      if (downloader == null) {
        downloader = createDefaultDownloader();
      }
      return new DownloadManager(this);
    }
  }
}

File: library/src/main/kotlin/com/coolerfall/download/DownloadManager.kt
package com.coolerfall.download

import android.content.ContentValues
import android.content.Context
import android.net.Uri
import android.os.Build.VERSION
import android.os.Build.VERSION_CODES
import android.os.Environment
import android.provider.MediaStore.Downloads
import android.provider.MediaStore.Files.FileColumns
import android.webkit.MimeTypeMap
import com.coolerfall.download.DownloadState.INVALID
import com.coolerfall.download.Helper.copy
import com.coolerfall.download.Helper.createDefaultDownloader
import com.coolerfall.download.Helper.resolvePath
import com.coolerfall.download.Preconditions.checkNotNull
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStream

/**
 * A manager used to manage the downloading.
 *
 * @author Vincent Cheung (coolingfall@gmail.com)
 */
class DownloadManager internal constructor(builder: Builder) {
  private val context: Context
  private val downloader: Downloader
  private val threadPoolSize: Int
  private val logger: Logger
  private val downloadRequestQueue: DownloadRequestQueue
  private val rootDownloadDir: String

  init {
    context = checkNotNull(builder.context, "context == null").applicationContext
    downloader = checkNotNull(builder.downloader, "downloader == null")
    threadPoolSize = builder.threadPoolSize
    logger = builder.logger
    downloadRequestQueue = DownloadRequestQueue(threadPoolSize, logger)
    downloadRequestQueue.start()
    val downloadDir = context.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS)
    checkNotNull(downloadDir, "shared storage is not currently available")
    rootDownloadDir = downloadDir?.absolutePath ?: ""
  }

  /**
   * Add one download request into the queue.
   *
   * @param request download request
   * @return download id, if the id is not set, then manager will generate one.
   * if the request is in downloading, then -1 will be returned
   */
  fun add(request: DownloadRequest): Int {
    checkNotNull(request, "request == null")
    if (isDownloading(
            request.uri()
                .toString()
        )
    ) {
      return -1
    }
    request.rootDownloadDir(rootDownloadDir)
    request.downloader(downloader.copy())

    /* add download request into download request queue */
    return if (downloadRequestQueue.add(request)) request.downloadId() else -1
  }

  /**
   * Query download from download request queue.
   *
   * @param downloadId download id
   * @return download state
   */
  fun query(downloadId: Int): DownloadState {
    return downloadRequestQueue.query(downloadId)
  }

  /**
   * Query download from download request queue.
   *
   * @param url download url
   * @return download state
   */
  fun query(url: String): DownloadState {
    return downloadRequestQueue.query(Uri.parse(url))
  }

  /**
   * To check if the download was in the request queue.
   *
   * @param downloadId downalod id
   * @return true if was downloading, otherwise return false
   */
  fun isDownloading(downloadId: Int): Boolean {
    return query(downloadId) !== INVALID
  }

  /**
   * To check if the download was in the request queue.
   *
   * @param url downalod url
   * @return true if was downloading, otherwise return false
   */
  fun isDownloading(url: String): Boolean {
    return query(url) !== INVALID
  }

  /**
   * Get the download task size.
   *
   * @return the task size
   */
  val taskSize: Int
    get() = downloadRequestQueue.downloadingSize

  /**
   * Cancel the download according to download id.
   *
   * @param downloadId download id
   * @return true if download has canceled, otherwise return false
   */
  fun cancel(downloadId: Int): Boolean {
    return downloadRequestQueue.cancel(downloadId)
  }

  /**
   * Cancel all the downloading in queue.
   */
  fun cancelAll() {
    downloadRequestQueue.cancelAll()
  }

  /**
   * Release all the resource.
   */
  fun release() {
    downloadRequestQueue.release()
  }

  /**
   * Copy downloaded file to external public download directory.
   *
   * @param filepath filepath of downloaded file
   * @return true if copy successfully, otherwise return false
   */
  fun copyToPublicDownloadDir(filepath: String): Boolean {
    if (!filepath.startsWith(rootDownloadDir)) {
      logger.log("Only files of current app can be exported")
      return false
    }

    try {
      openOutputStream(filepath)?.use { it ->
        val fis = FileInputStream(filepath)
        copy(fis, it)
        return true
      }
    } catch (e: Exception) {
      logger.log("Failed to copy file to public download directory: " + e.message)
    }

    return false
  }

  @Suppress("DEPRECATION")
  @Throws(IOException::class) private fun openOutputStream(
    filepath: String
  ): OutputStream? {
    val filename = filepath.substring(filepath.lastIndexOf(File.separator) + 1)

    return if (VERSION.SDK_INT >= VERSION_CODES.Q) {
      val contentValues = ContentValues()
      contentValues.put(
          FileColumns.RELATIVE_PATH,
          Environment.DIRECTORY_DOWNLOADS
      )
      contentValues.put(FileColumns.DISPLAY_NAME, filename)
      val index = filename.lastIndexOf(".")
      if (index > 0) {
        val mimeType = MimeTypeMap.getSingleton()
            .getMimeTypeFromExtension(filename.substring(index + 1))
        contentValues.put(FileColumns.MIME_TYPE, mimeType)
      }
      val contentResolver = context.contentResolver
      val uri = contentResolver.insert(Downloads.EXTERNAL_CONTENT_URI, contentValues)
          ?: throw IOException("Cannot get shared download directory")
      contentResolver.openOutputStream(uri)
    } else {
      val dir =
        Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).absolutePath
      val outputFilepath = resolvePath(dir, filepath.substring(rootDownloadDir.length + 1))
      FileOutputStream(outputFilepath)
    }
  }

  fun newBuilder(): Builder {
    return Builder(this)
  }

  class Builder {
    var context: Context? = null
    var downloader: Downloader? = null
    var threadPoolSize: Int
    var logger: Logger

    constructor() {
      threadPoolSize = 3
      logger = Logger.EMPTY
    }

    internal constructor(downloadManager: DownloadManager) {
      context = downloadManager.context
      downloader = downloadManager.downloader
      threadPoolSize = downloadManager.threadPoolSize
      logger = downloadManager.logger
    }

    fun context(context: Context): Builder {
      this.context = context
      return this
    }

    fun downloader(downloader: Downloader): Builder {
      this.downloader = downloader
      return this
    }

    fun threadPoolSize(threadPoolSize: Int): Builder {
      this.threadPoolSize = threadPoolSize
      return this
    }

    fun logger(logger: Logger): Builder {
      this.logger = logger
      return this
    }

    fun build(): DownloadManager {
      if (downloader == null) {
        downloader = createDefaultDownloader()
      }
      return DownloadManager(this)
    }
  }
}